lib/std/compress/zlib.zig

```diff
@@ -84,14 +84,101 @@ pub fn ZlibStream(comptime ReaderType: type) type {
     };
 }
 
-pub fn zlibStream(allocator: mem.Allocator, reader: anytype) !ZlibStream(@TypeOf(reader)) {
-    return ZlibStream(@TypeOf(reader)).init(allocator, reader);
+pub fn zlibStreamReader(allocator: mem.Allocator, reader: anytype) !ZlibStreamReader(@TypeOf(reader)) {
+    return ZlibStreamReader(@TypeOf(reader)).init(allocator, reader);
+}
+
+pub const CompressionLevel = enum(u2) {
+    no_compression = 0,
+    fastest = 1,
+    default = 2,
+    maximum = 3,
+};
+
+pub const CompressionOptions = struct {
+    level: CompressionLevel = .default,
+};
+
+pub fn ZlibStreamWriter(comptime WriterType: type) type {
+    return struct {
+        const Self = @This();
+
+        const Error = WriterType.Error ||
+            deflate.Compressor(WriterType).Error;
+        pub const Writer = io.Writer(*Self, Error, write);
+
+        allocator: mem.Allocator,
+        deflator: deflate.Compressor(WriterType),
+        in_writer: WriterType,
+        hasher: std.hash.Adler32,
+
+        fn init(allocator: mem.Allocator, dest: WriterType, options: CompressionOptions) !Self {
+            // Zlib header format is specified in RFC1950
+            const CM: u4 = 8; // DEFLATE
+            const CINFO: u4 = 7; // 32K window
+            const CMF: u8 = (@as(u8, CINFO) << 4) | CM;
+
+            const FLEVEL: u2 = @enumToInt(options.level);
+            const FDICT: u1 = 0; // No preset dictionary support
+            const FCHECK: u5 = 28;
+            const FLG = (@as(u8, FLEVEL) << 6) | (@as(u8, FDICT) << 5) | FCHECK;
+
+            const compression_level: deflate.Compression = switch (options.level) {
+                .no_compression => .no_compression,
+                .fastest => .best_speed,
+                .default => .default_compression,
+                .maximum => .best_compression,
+            };
+
+            try dest.writeAll(&.{ CMF, FLG });
+
+            return Self{
+                .allocator = allocator,
+                .deflator = try deflate.compressor(allocator, dest, .{ .level = compression_level }),
+                .in_writer = dest,
+                .hasher = std.hash.Adler32.init(),
+            };
+        }
+
+        pub fn write(self: *Self, bytes: []const u8) Error!usize {
+            if (bytes.len == 0) {
+                return 0;
+            }
+
+            const w = try self.deflator.write(bytes);
+
+            if (w != 0) {
+                self.hasher.update(bytes[0..w]);
+                return w;
+            }
+
+            return 0;
+        }
+
+        pub fn writer(self: *Self) Writer {
+            return .{ .context = self };
+        }
+
+        pub fn deinit(self: *Self) void {
+            self.deflator.deinit();
+        }
+
+        pub fn close(self: *Self) !void {
```

Comment:
The name `close` is taken from the underlying deflate compressor, this seems to be an already accepted pattern in the compressor so I don't see the problem with using it on the `zlibStreamWriter` which is just a thin layer around it

Reply:
This is a problem, because Zig tries to make things consistent for the user and one of that consistency things is that we expect `close()` and `deinit()` methods must never fail. You can try `fn close lib/std/` or `grep -r close lib/std` yourself to see, if you can find a `close` method which returns an error.

This allows the user to rely on
```zig
var thing = Mything.init()
defer thing.close()
```
to just work on first attempt with a completer without cognitive overhead.
If it doesnt complete, then one knows there is special handling.
If `close` would fail, this would break the users workflow.